### PR TITLE
Various fixes to the partitioning system

### LINKED ISSUE #548: Soundness bug in parallel opensmt (SMTS)
In some cases the parallel opensmt (SMTS) implementation reports satisfiable instances as unsatisfiable.  The sequential solver is not affected by this bug.  At the moment it seems that this is coming from the code enabled with `PARALLEL=ON` in opensmt.  Auxiliary variables being communicated between the solvers can sometimes be mapped to incorrect terms.

### LINKED ISSUE #550: Incorrect implementation of the scattering algorithm
Currently opensmt's splitting system implements the scattering algorithm incorrectly.

The idea in scattering is, given an instance F, construct new partitions F_1, ..., F_n such that (1) F_i /\ F_j is unsatisfiable whenever i != j, and (2) F_1 \/ ... \/ Fn is logically equivalent to F.  The current implementation does not satisfy (1).

While the current implementation of the scattering algorithm is not unsound, it most likely performs less well than the intended implementation.

### PR DESCRIPTION
**Pull request by aehyvari**
This PR contains three improvements to opensmt's partitioning system:
- The correct implementation of scattering (see #550)
- A probable fix for the correctness issue in SMTS (see #548)
- A performance improvement related to constructing splits and sharing clauses.

In addition the PR introduces internal checks for the correctness of the partitions in debug mode and improvements for the partition system's regression tests.

After serialising opensmt's PTRefs to smt-lib syntax, even when expressed using let-statements, the representation may grow extremely large.  This is likely due to the Tseitin expansion, as it is possible that the whole original instance is represented by a single literal which then needs to be rewritten in the smt-lib syntax.  The solution suggested by this PR and some previous ones is to limit the number of nodes in the PTRefs that can be shared.  This affects the possible partitions as well as the lemmas to be shared, but the overall effect on performance seems to be positive.  In the current implementation, when ran on smt-lib benchmarks, it seems that a limit of 100 nodes per PTRef is a reasonable compromise.

Fixes #550.
Fixes #548.

### CODE REVIEW

**Review by blishko — COMMENTED** [review]
Seems good to me.
I have not seen any problem, although I am not very familiar with this code.
Left just two small comments.

File: src/parallel/MainSplitter.cc [hunk 1]
```diff
@@ -91,24 +91,77 @@ std::unique_ptr<SimpSMTSolver> MainSplitter::createInnerSolver(SMTConfig & confi
...
+    }
+
+    std::vector<std::string> partitions;
+    for (PTRef trWithAuxVars : partitionsTr) {
+        PTRef tr = logic.removeAuxVars(trWithAuxVars);
+        partitions.push_back(logic.dumpWithLets(tr));
     }
+
+    assert(
+        [this](vec<PTRef> const & partitions) {
```
[the remark this answers is not in the document]
Comment (aehyvari) [reply src/parallel/MainSplitter.cc:111]: Sound good!

**Review by MasoudAsadzade — COMMENTED** [review]
Following Plots will confirm that this fix doesn't have overhead while learning clauses. and that MAX_SIZE of 100 is the best.
non-incremental-QF-LRA-splittNonLemmaShareMS100-2022-08-24_vs_splittLemmaShareMS100-2022-08-24

cactus-plot-resultdir-shellTime
cactus-plot-resultdir-100To1000

**Comment by MasoudAsadzade** [comment]
Looks good!

File: src/parallel/LookaheadSplitter.cc [hunk 2]
```diff
@@ -36,7 +36,7 @@ bool LookaheadSplitter::createSplitLookahead(LASplitNode & n)
             // duplicated.  Do not report duplicates.
             tmp.push(l);
             printf("%s%d ", sign(l) ? "-" : "", var(l));
-            sd.addConstraint(tmp);
+            sd.addConstraint(std::move(tmp));
```
[the remark this answers is not in the document]
Comment (aehyvari) [reply src/parallel/LookaheadSplitter.cc]: Fixed.

File: src/parallel/MainSplitter.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (aehyvari) [reply src/parallel/MainSplitter.cc]: Sounds good!

File: src/parallel/LookaheadSplitter.cc [hunk 3]
```diff
@@ -34,9 +34,8 @@ bool LookaheadSplitter::createSplitLookahead(LASplitNode & n)
             // In cases where the LA solver couldn't propagate due to
             // literal being already assigned, the literal may be
             // duplicated.  Do not report duplicates.
-            tmp.push(l);
```
Comment (blishko) [inline src/parallel/LookaheadSplitter.cc:37]: There is now an unused `vec<Lit> tmp` on line 31.
I wonder if the compiler will warn about it.

**Review by blishko — APPROVED** [review]
LGTM!

File: src/parallel/LookaheadSplitter.cc (hunk 3 above)
Comment (aehyvari) [reply src/parallel/LookaheadSplitter.cc:37]: It doesn't, but I'll fix it.

**Pull request merged**